gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/user/impl/UserStoreImplV2.java

```diff
@@ -0,0 +1,90 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.gravitee.am.gateway.handler.common.user.impl;
+
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.gravitee.am.gateway.handler.common.user.UserStore;
+import io.gravitee.am.gateway.handler.common.user.UserValueMapper;
+import io.gravitee.am.model.User;
+import io.gravitee.node.api.cache.Cache;
+import io.gravitee.node.api.cache.CacheConfiguration;
+import io.gravitee.node.api.cache.CacheManager;
+import io.reactivex.rxjava3.core.Completable;
+import io.reactivex.rxjava3.core.Maybe;
+import org.springframework.core.env.Environment;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @author Eric LELEU (eric.leleu at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class UserStoreImplV2 implements UserStore {
+
+    private static final String SEPARATOR = ":";
+
+    private Cache<String, User> idCache;
+    private Cache<String, User> gisCache;
+
+    private int ttl;
+
+    public UserStoreImplV2(CacheManager cacheManager, Environment environment) {
+        final var mapper = new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        final var valueMapper = new UserValueMapper(mapper);
+        this.idCache = cacheManager.getOrCreateCache("userStoreById", valueMapper);
+        this.gisCache = cacheManager.getOrCreateCache("userStoreByGis", valueMapper);
+        this.ttl = environment.getProperty("http.cookie.session.cache.ttl", Integer.class, 36000);
+    }
+
+    @Override
+    public Maybe<User> add(User user) {
+        return idCache.rxPut(user.getId(), user, ttl, TimeUnit.SECONDS)
+                .concatWith(gisCache.rxPut(generateInternalSubFrom(user.getSource(), user.getExternalId()), user, ttl, TimeUnit.SECONDS))
+                .lastElement();
+    }
+
+    @Override
+    public Completable remove(String userId) {
+        return idCache.rxEvict(userId).flatMap(user -> gisCache.rxEvict(generateInternalSubFrom(user.getSource(), user.getExternalId()))).ignoreElement();
+    }
+
+    @Override
+    public Maybe<User> get(String userId) {
+        return idCache.rxGet(userId);
+    }
+
+    @Override
+    public Completable removeByInternalSub(String gis) {
+        return gisCache.rxEvict(gis).flatMap(user -> idCache.rxEvict(user.getId())).ignoreElement();
+    }
+
+    @Override
+    public Maybe<User> getByInternalSub(String gis) {
+        return gisCache.rxGet(gis);
+    }
+
+    @Override
+    public Completable clear() {
+        return gisCache.rxClear().andThen(idCache.rxClear());
+    }
+
+    private String generateInternalSubFrom(String src, String externalId) {
```

Comment:
Suggestion. This is done here and in UserServiceImplV2. Maybe you can extract this to some utils, or at least keep SEPARATOR variable in one place? Currently there are 3 places where SEPARATOR is defined.